ledger/simulation/simulator.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package simulation
+
+import (
+	"errors"
+
+	"github.com/algorand/go-algorand/data"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/verify"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+)
+
+// ==============================
+// > Simulator Ledger
+// ==============================
+
+// simulatorLedger patches the ledger interface to use a constant latest round.
+type simulatorLedger struct {
+	*data.Ledger
+	start basics.Round
+}
+
+// Latest is part of the LedgerForSimulator interface.
+// We override this to use the set latest to prevent racing with the network
+func (l simulatorLedger) Latest() basics.Round {
+	return l.start
+}
+
+// LookupLatest would implicitly use the latest round in the _underlying_
+// Ledger, it would give wrong results if that ledger has moved forward. But it
+// should never be called, as the REST API is the only code using this function,
+// and the REST API should never have access to a simulatorLedger.
+func (l simulatorLedger) LookupLatest(addr basics.Address) (basics.AccountData, basics.Round, basics.MicroAlgos, error) {
+	err := errors.New("unexpected call to LookupLatest")
+	return basics.AccountData{}, 0, basics.MicroAlgos{}, err
+}
+
+// ==============================
+// > Simulator Errors
+// ==============================
+
+// SimulatorError is the base error type for all simulator errors.
+type SimulatorError struct {
+	err error
+}
+
+func (s SimulatorError) Error() string {
+	return s.err.Error()
+}
+
+func (s SimulatorError) Unwrap() error {
+	return s.err
+}
```

Comment:
I can't figure out why the type `SimulatorError` exists. Nobody ever tests for it.  At the very least it does not seem to need to be public, but I can't see why ever use of it can be "unwrapped" so:
`EvalFailureError{SimulatorError{err}}` is `EvalFailureError{err}`

Reply:
Perhaps SimulatorError should be `basics.WrappedError`. It seems like your using it as a kind of mix-in to give wrapping functionality to those other two error types?

Reply:
I wrote it more as a utility for anyone using the Simulator since it's an issue I ran into with checking what type an error is within transaction evaluation. For example, if another function one-level above the `Simulate` method has multiple types of errors, it would be nice to have a "catch-all" way of asking if it was _any_ type of simulator error.

Or put another way, it's nice to have options for being both specific (with `errors.As(err, &evalFailureErr)`) and general (with `errors.As(err, &simulatorError)`).